e2e/testcases/notification_test.go

```diff
@@ -238,10 +251,206 @@ func TestNotification(t *testing.T) {
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+	// first notification error includes a warning of known host added which includes an IP address.
+	// the IP address is dynamic, so skip checking the message for the first notification
+	require.Equal(t, rootSyncNotificationCredentialHash, records.Records[0].Auth)
+	// subsequent git-sync iterations do not include the warning, but otherwise should
+	// surface a stable error message.
+	require.Equal(t, rootSyncNotificationCredentialHash, records.Records[1].Auth)
+	// note the commit is empty because the branch does not exist
+	require.Equal(t, `{
+  "content": {
+    "raw": "RootSync root-sync-2 failed to sync commit  on branch nonexistent-branch!\n\nKNV2004: error in the git-sync container: {\"Msg\":\"error syncing repo, will retry\",\"Err\":\"Run(git clone -v --no-checkout -b nonexistent-branch --depth 1 git@test-git-server.config-management-system-test:/git-server/repos/config-management-system/root-sync-2 /repo/source): exit status 128: { stdout: \\\"\\\", stderr: \\\"Cloning into \'/repo/source\'...\\\\nwarning: Could not find remote branch nonexistent-branch to clone.\\\\nfatal: Remote branch nonexistent-branch not found in upstream origin\\\\nfatal: The remote end hung up unexpectedly\\\" }\"}\u000A\u000AFor more information, see https://g.co/cloud/acm-errors#knv2004\n\n"
+  }
+}`, records.Records[1].Message)
+	// creating a new error should produce another notification
+	rootSync.Spec.Git.Branch = "another-nonexistent-branch"
+	nt.RootRepos[configsync.RootSyncName].Add(nomostest.StructuredNSPath(rootSyncNN.Namespace, rootSyncNN.Name), rootSync)
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush(fmt.Sprintf("Switch RootSync %s to another-nonexistent-branch", rootSyncNN.Name))
+	nt.WaitForRootSyncSourceError(rootSyncNN.Name,
+		status.SourceErrorCode,
+		"Remote branch another-nonexistent-branch not found in upstream")
+	// query RootSync to get the generation
+	rootSyncObj = &v1beta1.RootSync{}
+	if err := nt.KubeClient.Get(rootSyncNN.Name, rootSyncNN.Namespace, rootSyncObj); err != nil {
+		nt.T.Fatal(err)
+	}
+	err = nt.Watcher.WatchObject(kinds.NotificationV1Beta1(), rootSyncNN.Name, rootSyncNN.Namespace,
+		[]testpredicates.Predicate{testpredicates.NotificationHasStatus(v1beta1.NotificationStatus{
+			ObservedGeneration: rootSyncObj.Generation,
+			Commit:             "", // commit is empty, because the Syncing condition does not have the commit
+			Deliveries: []v1beta1.NotificationDelivery{
+				{
+					Trigger:         "on-sync-failed",
+					Service:         "local",
+					Recipient:       "",
+					AlreadyNotified: true,
+				},
+			},
+		})},
+	)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+	// notification from first branch failure, including warning of new known host
+	// notification from first branch failure
+	// notification of second branch failure, including warning of new known host
+	// notification of second branch failure
+	records, err = waitForNotifications(nt, 4)
```

Comment:
I was hoping my comments above would explain it, but I'll try to provide the context here:

There are two notifications each time the git-sync failure occurs. That's because the first iteration of the git-sync container includes a warning in the error about the known host being added. After the first iteration that warning is no longer included, so the error changes after the first iteration. The error message is otherwise stable, but this means there are two notifications sent for the error.

I added logging to the test, so you can see an example here:
https://oss.gprow.dev/view/gs/oss-prow-build-kpt-config-sync/pr-logs/pull/GoogleContainerTools_kpt-config-sync/542/kpt-config-sync-presubmit-e2e-multi-repo-test-group2/1648186128898985984#1:build-log.txt%3A8781

Reply:
Note this is because each time we change the branch it means the git-sync container is restarted to receive the new branch variable, so it needs to add the known host each time

Reply:
That makes sense.